### PR TITLE
Add commands to get index of booted, pending and previous deployments

### LINKED ISSUE #3142: Alias to pin current deployment
It would be good if there was a quick and easy way to pin the current deployment without needing to rely on multiple commands: 
- `rpm-ostree status -v` - to find out which index the current deployed system is
  - It changes from 0 to 1 when there's an update staged
- `ostree admin pin <index>` - to actually pin the deployment

It would be great if there was a simple alias called `current` (or whatever you feel makes the most sense) which points to the deployed index.
This could also be extended to include keywords like `previous` or `staged`, but I see less reason for this to be used.

Therefore, the user could run `ostree admin pin current` and be done with it.

### PR DESCRIPTION
**Pull request by p5**
Closes #3142 

Right now, it's difficult to pin the current deployment without needing to look into the output of some other tooling (like rpm-ostree) to get the index of each deployment.  This index also is not consistent - the current deployment could be 0 when you first boot the system then 1 shortly after. 

This change makes it easy to get the index of the current or future deployment so it can be piped into other commands (such as ostree admin pin).

(Please note: I have not yet built or tested this PR - I am working on figuring out how to compile it)

Huge thanks to Eric for authoring this patch for me!

TODO before merge:
- [ ] Squash all commits into one
- [ ] Get confirmation for flag descriptions & man pages
- [x] Confirm the new functionality works locally

Co-authored-by: Eric Curtin <ecurtin@redhat.com>

### CODE REVIEW

File: src/ostree/ot-admin-builtin-status.c
Comment (ericcurtin) [inline src/ostree/ot-admin-builtin-status.c:48]: So the:

"Output \"default\" if booted into the default deployment, otherwise \"not-default\""

shoudl have something that actually makes sense for these options

Comment (p5) [reply src/ostree/ot-admin-builtin-status.c]: Apologies - should have picked up on this :facepalm: 

Have added some descriptors that make sense now, but let me know if you want these changing.

Have squashed my commits into one as most projects ask for this anyway, therefore cannot point you to the commit with this fix.

**Comment by ericcurtin** [comment]
Will need equivalent man page updates too.

You can also collapse those 3 else if's into ingel return statements if you write a function that takes two parameters.

I'm in bed now but I can write the function signature in the morning to point you in the right direction.

**Comment by ericcurtin** [comment]
Noticed your comment about how to build... I use this:

https://github.com/ericcurtin/staging/blob/master/build.sh

**Comment by p5** [comment]
> Will need equivalent man page updates too.
> 
> You can also collapse those 3 else if's into ingel return statements if you write a function that takes two parameters.

I believe I have done this:

- Man pages: [1fe36d3](https://github.com/ostreedev/ostree/pull/3144/commits/1fe36d3430652579207eec55791d3925bd52438b)
- Refactoring: [114cd15](https://github.com/ostreedev/ostree/pull/3144/commits/114cd158d90746898b31720addec5b169cb5ab88)

I will mark this as draft for now as there's a high chance I have messed something up and want to test this out locally.... Tomorrow...

But please let me know if the refactoring is not what you intended.

**Comment by p5** [comment]
Ah, just found you already upload built binaries during CI, so that makes testing these changes locally so much easier!
I believe everything is working as expected!

Command outputs: 

```
☁ bin  ./ostree admin status -b    
1
☁  bin  ./ostree admin status -p
0
☁  bin  ./ostree admin status -r
2
```

RPM Ostree Status:

<details>

```
State: idle
AutomaticUpdates: stage; rpm-ostreed-automatic.timer: last run 15h ago
Deployments:
  ostree-image-signed:registry:ghcr.io/rsturla/eternal-linux/lumina:39-nvidia (index: 0)
                   Digest: sha256:a3f55eb777e88c71a692e655c838fd8e6b48186f8422d2630b96bea85a09e71a
                  Version: 39.20240121.0 (2024-01-22T02:11:23Z)
               BaseCommit: 345d49ead85e7f6b14ae171861f8e66c65c3bad8a79a9e8c90980c0d139e3c0d
                   Commit: a7647d086bf0263fa0a56497d4a418fd3b6a29c6219fe30a88a279449f7a8528
                           ├─ 1password (2023-12-30T00:40:06Z)
                           ├─ fedora (2023-11-01T00:12:39Z)
                           ├─ fedora-cisco-openh264 (2023-03-14T10:57:01Z)
                           ├─ fedora-cisco-openh264-debuginfo (2023-03-14T10:57:01Z)
                           ├─ google-chrome (2024-01-18T15:42:25Z)
                           ├─ updates (2024-01-22T01:06:29Z)
                           └─ updates-archive (2024-01-22T01:30:46Z)
                   Staged: yes
                StateRoot: fedora
          LayeredPackages: 1password google-chrome-stable

● ostree-image-signed:registry:ghcr.io/rsturla/eternal-linux/lumina:39-nvidia (index: 1)
                   Digest: sha256:815aa0b03438ca450484a8ef2e3c63ec4b0fae41a36836c80e6e8837524e0fe6
                  Version: 39.20240121.0 (2024-01-21T13:26:40Z)
               BaseCommit: b07f32f84ed89fa31bf3765be9f880edf316298e4a626518d87910679e4a230e
                   Commit: a31e463b9501a09a431a1f25c455fb5299d45d5e92486eb54ef1db411458e904
                           ├─ 1password (2023-12-30T00:40:06Z)
                           ├─ fedora (2023-11-01T00:12:39Z)
                           ├─ fedora-cisco-openh264 (2023-03-14T10:57:01Z)
                           ├─ fedora-cisco-openh264-debuginfo (2023-03-14T10:57:01Z)
                           ├─ google-chrome (2024-01-18T15:42:25Z)
                           ├─ updates (2024-01-21T03:22:03Z)
                           └─ updates-archive (2024-01-21T03:46:07Z)
                StateRoot: fedora
          LayeredPackages: 1password google-chrome-stable

  ostree-image-signed:registry:ghcr.io/rsturla/eternal-linux/lumina:39-nvidia (index: 2)
                   Digest: sha256:8ea385f222b1fdaf8aeebd30ade54c33db457e20b2539c4320b6eeae82815472
                  Version: 39.20240120.0 (2024-01-21T01:48:06Z)
               BaseCommit: 01c6ea8c5e991f74c001c55b4360a4355ff044faafe2f4229b7a5c2859217934
                   Commit: e926c5f3a8eea670eb5ec2d983431de8fe201cabf6c89fef9b817584dda110e9
                           ├─ 1password (2023-12-30T00:40:06Z)
                           ├─ fedora (2023-11-01T00:12:39Z)
                           ├─ fedora-cisco-openh264 (2023-03-14T10:57:01Z)
                           ├─ fedora-cisco-openh264-debuginfo (2023-03-14T10:57:01Z)
                           ├─ google-chrome (2024-01-18T15:42:25Z)
                           ├─ updates (2024-01-20T03:14:28Z)
                           └─ updates-archive (2024-01-20T03:39:19Z)
                StateRoot: fedora
          LayeredPackages: 1password google-chrome-stable

  ostree-image-signed:registry:ghcr.io/rsturla/eternal-linux/lumina:39-nvidia (index: 3)
                   Digest: sha256:0eb71fc0fedc315224dcc0facbab0889a1002582a4f92fa8c399163f6e0e2bf2
                  Version: 39.20240118.0 (2024-01-18T13:31:30Z)
               BaseCommit: e9e6740ab9998898792320d0ff773c03f38e232ccf2bf64314422260a293c420
                   Commit: c959c74c35711466eed42789a9c1d297caa0c9bc2840bd94e254e6e5e70bec75
                           ├─ 1password (2023-12-30T00:40:06Z)
                           ├─ fedora (2023-11-01T00:12:39Z)
                           ├─ fedora-cisco-openh264 (2023-03-14T10:57:01Z)
                           ├─ fedora-cisco-openh264-debuginfo (2023-03-14T10:57:01Z)
                           ├─ google-chrome (2024-01-16T18:03:56Z)
                           ├─ rpmfusion-free (2023-11-04T16:49:08Z)
                           ├─ rpmfusion-free-updates (2024-01-12T08:11:53Z)
                           ├─ rpmfusion-nonfree (2023-11-04T17:26:32Z)
                           ├─ rpmfusion-nonfree-updates (2024-01-12T08:27:46Z)
                           ├─ updates (2024-01-18T01:40:54Z)
                           └─ updates-archive (2024-01-18T02:04:23Z)
                StateRoot: fedora
          LayeredPackages: 1password google-chrome-stable
                   Pinned: yes
```

</details>

**Comment by ericcurtin** [comment]
After you build locally, you can just test from the git repo using commands like:

`LD_LIBRARY_PATH=.libs ./.libs/ostree admin status -r`

I pulled the branch down, de-duplicated the code further, made some other styling changes to be consistent with the rest of the project (we don't do camelCase for example).

I squashed the commits. Fixed up the commit message (we put "status: " for example in the commit message when we change this part). The angular brackets were missing in the Co-authored-by for example, so basically, if you push this, we should be pretty good:

https://github.com/ericcurtin/ostree/commit/34a4e187fb3e1170ac525df568d3c2555d0ce9cc

**Comment by ericcurtin** [comment]
Also had to remove changes accidentally made to libglnx, that may have been my mistake, unsure...

**Comment by p5** [comment]
> I squashed the commits. Fixed up the commit message (we put "status: " for example in the commit message when we change this part). The angular brackets were missing in the Co-authored-by for example, if we push this, we should be pretty good:
https://github.com/ericcurtin/ostree/commit/f892522028eb8ec61bf7e5464c64c41f28a7461c

All changes have been made.  

I'm not sure where "changes accidentally made to libglnx" could be found in your commit, so if I have missed something, please let me know.

And sorry it's taken so many revisions to get something that's in a good enough state.  I'm not familiar at all with C so it was very much trial and error to get something that seemed right.

**Comment by ericcurtin** [comment]
> > I squashed the commits. Fixed up the commit message (we put "status: " for example in the commit message when we change this part). The angular brackets were missing in the Co-authored-by for example, if we push this, we should be pretty good:
> > [ericcurtin@f892522](https://github.com/ericcurtin/ostree/commit/f892522028eb8ec61bf7e5464c64c41f28a7461c)
> 
> All changes have been made.
> 
> I'm not sure where "changes accidentally made to libglnx" could be found in your commit, so if I have missed something, please let me know.
> 
> And sorry it's taken so many revisions to get something that's in a good enough state. I'm not familiar at all with C so it was very much trial and error to get something that seemed right and CI didn't shout at.

Well you did a pretty amazing job at it for someone that doesn't no C! This LGTM, once Draft is removed I think we can merge unless another ostree maintainer wants to look at this also in the next 24 hours.

**Comment by ericcurtin** [comment]
As an aside, I'm looking for someone in the community with cycles and an Apple Silicon device to look at this:

https://github.com/ericcurtin/fedora-asahi-ostree-desktops/issues/1

if you know of someone that wants to look at this let me know.

It's hard for me to justify looking at it, as my Apple Silicon device is my daily driver and it blocks my Automotive work when I start flashing and rebooting it to test Asahi Kinoite/Silverblue.

**Review by ericcurtin — APPROVED** [review]

**Pull request closed without merging**
